### PR TITLE
Dikla Rikovitch Scrabble Octo

### PR DESCRIPTION
**Pull request by diklaaharoni**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the advantages of using git when collaboratively working on one code base? |you can work remotely with your partner  
Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? | because there is no need to create an instance of score 
Describe an example of a test you wrote for a _nominal case_. |
Describe an example of a test you wrote for an _edge case_. |
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | no
What was one method you and your pair used to debug code? | we used pry
Is there a specific piece of code you'd like feedback on? | all
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? |

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -1,9 +1,91 @@
...
+        return nil
+      elsif scrabble_word.length == 7
+        total_score += 50
+      end
+
+      scrabble_word.each do |letter|
+        case letter.downcase
+        when "a", "e", "i", "o", "u", "l", "n", "r", "s", "t"
+          total_score += 1
+        when "d", "g"
```
Comment (droberts-sea) [inline lib/scoring.rb:24]: Storing these as options in a `case` statement certainly works, but it is a single-use solution. If you ever needed to use letter values somewhere else, you'd have to copy paste. An alternative would be to build a data structure stored in a variable that maps letters to scores, something like this:

```ruby
LETTERS = {
  "A" => 1
  "B" => 3,
  "C" => 3,
  "D" => 2,
  # ...
}
```

Then to get the score for a letter, you can say `LETTERS[letter]`.

File: specs/scoring_spec.rb [hunk 2]
```diff
@@ -42,21 +42,27 @@
...
     it 'returns nil if no words were passed' do
+      Scrabble::Scoring.highest_score_from([]).must_be_nil
     end
 
     it 'returns the only word in a length-1 array' do
+      Scrabble::Scoring.highest_score_from(["apple"]).must_equal "apple"
     end
 
     it 'returns the highest word if there are two words' do
+      Scrabble::Scoring.highest_score_from(["apple", "pear"]).must_equal "apple"
```
Comment (droberts-sea) [inline specs/scoring_spec.rb:53]: Like we discussed in class, it would probably be wise to check multiple orderings of these words, that is, both `['apple', 'pear']` and `['pear', 'apple']`. You'd be surprised how often this sort of thing turns up a bug.

File: lib/player.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      @total_score = 0
+<<<<<<< HEAD
+=======
+      @tiles = []
+>>>>>>> 63e10b980789cec14cd9c17435143fee723b3388
+    end
+
+    def plays
+      return @player_array
+    end
```
Comment (droberts-sea) [inline lib/player.rb:25]: If you named the array `@plays`, you could make this method an `attr_reader`.

File: lib/player.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      @player_array << word
+      score = Scrabble::Scoring.score(word)
+      @total_score += score.to_i
+      return score
+    end
+
+    def won?
+      if @total_score > 100
+        return true
+      else
```
Comment (droberts-sea) [inline lib/player.rb:40]: Since `>` will always return `true` or `false`, this could be shortened to:

```ruby
def won?
  return total_score > 100
end
```

File: specs/player_spec.rb [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+      word = "dikla"
+      user = Scrabble::Player.new(word)
+      scrabble_word = user.plays
+      scrabble_word.must_be_kind_of Array
+    end
+  end
+
+  describe 'play' do
+    it 'Adds the input word to the plays Array' do
+
```
Comment (droberts-sea) [inline specs/player_spec.rb:32]: What happens if the player has already won? Is the word recorded? What does it return?

File: specs/player_spec.rb [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+      end
+      # Act
+      result += user.play("dog")
+      # Assert
+      result.must_equal 15
+    end
+  end
+
+  describe 'won?' do
+    it 'Returns true if the player has over 100 points' do
```
Comment (droberts-sea) [inline specs/player_spec.rb:82]: I would be interested in investigating a little closer around 100 points, to more clearly define the "edge" of the behavior. In particular, what happens when the player has:

- 99 points
- 100 points
- 101 points

File: lib/tilebag.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      "p" => 2, "q" => 1, "r" => 6, "s" => 4, "t" => 6,
+      "u" => 4, "v" => 2, "w" => 2, "x" => 1, "y" => 2, "z" => 1
+      }
+
+
+    def initialize
+      @tiles = []
+      DEFAULT_TILES.each do |letter, count|
+        count.times {@tiles << letter}
+      end
```
Comment (droberts-sea) [inline lib/tilebag.rb:22]: I love this strategy for getting the initial tiles into the tile bag! Keeping them in an array is much simpler than trying to manage them in a hash, for example.

File: specs/tilebag_spec.rb [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+      result.length.must_equal 5
+    end
+
+    it "removes the tiles from the default set " do
+      tile_bag = Scrabble::TileBag.new
+      before_count = 98
+      tile_bag.draw_tiles(5)
+      after_count = tile_bag.tiles_remaining
+      after_count.must_equal 93
+    end
```
Comment (droberts-sea) [inline specs/tilebag_spec.rb:64]: Other questions I would like your tests to answer:

- Am I able to draw the last tile from the bag?
- What happens if I try to draw more tiles than are currently in the bag?
- What if I try to draw 0 or -3 tiles?

**Comment by droberts-sea** [comment]
# Scrabble
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | Several of these answers are missing, and the ones you do have are pretty sparse. In future pair projects, please set aside some time with your partner to go over these questions together before the project is due.
Both Teammates contributed to the codebase | yes
Regular Commits with meaningful commit messages | I see a lot of commit messages that say something like "no changes" or "nothing to commit", but which are actually big commits full of work. In the future, please try to provide commit messages that describe how and why the code changed.<br><br>This submission is currently full of git merge-conflict markers! These need to be removed before you're really finished merging. To grade it, I rolled back to a previous version before all those came in.
**Scoring class** |
`score` method | yes
Uses appropriate data structure to store the letter score | see inline
Appropriately handles edge cases | yes
`highest_score_from` method | yes
Appropriately handles edge cases | yes - good work getting the tie-breaking logic
Test implementations for `highest_score_from`| mostly - see inline
**Player class** |
Uses an array to keep track of words played | yes
Uses existing code to create `highest_scoring_word` and `highest_word_score` | yes
Returns `true` or `false` from `won?` method | yes
Tests | missing some - see inline
**TileBag class** |
Uses a data structure to keep the set of default tiles in the bag | yes
`draw_tiles` method uses the parameter to draw a set of random tiles | yes
`tiles_remaining` method returns a count | yes
Tests `draw_tiles` edge cases | missing some - see inline

Good job overall! The big learning goals I'm looking to see demonstrated on this project are:

- Building complex logic (tie-breaking logic in `highest_score_from`)
- Test coverage
- Object composition (interaction between `Player` and `TileBag`)

I feel that you've done a good job of demonstrating the first and the third of these. Your test coverage could use some work - several of your methods are missing tests entirely, or are missing interesting cases. I've left some inline comments for you to review, please do so, and keep up the hard work!
